Cython/Compiler/TypeSlots.py

```diff
@@ -1102,20 +1106,32 @@ def get_special_method_signature(self, name):
 
     def get_slot_by_method_name(self, method_name):
         # For now, only search the type struct, no referenced sub-structs.
-        return self.method_name_to_slot[method_name]
+        return self._get_slot_by_method_name(method_name)
+
+    def __iter__(self):
+        # make it easier to iterate over all the plots
+        return iter(self.slot_table)
 
 
 _slot_table_dict = {}
+special_method_names = set()
 
 def get_slot_table(compiler_directives):
-    # use "get" here with a default since the builtin type classes don't have
-    # directives set
-    old_binops = compiler_directives.get('c_api_binop_methods', False)
+    if not compiler_directives:
```

Comment:
Hmm. I was wondering if the argument should then be optional, but that would make it too easy to forget in the normal case. I was also wondering if the truth value is the right thing to check here or if we should pass None instead for the explicit "don't have directives" case. I guess it doesn't matter – the directives should never be empty, unless explicitly passed empty. We might want to pass a keyword argument in the latter case, though, to make the call more readable … in which case None would also work as an argument… Not sure which is better.

Reply:
The main place this gets called without directives is when setting up the builtin classes (I think it's indirectly, from Builtin.py to Symtab.py"). In this case it gets an empty dictionary since the builtin classes don't have directives initialized on their scope. Therefore I think the function needs to be able to handle an empty dictionary (or change some other code more significantly).